webapp/components/channel_switch_modal.jsx

```diff
@@ -85,7 +91,16 @@ export default class SwitchChannelModal extends React.Component {
         const name = this.state.text.trim();
         let channel = null;
 
-        // TODO: Replace this hack with something reasonable
+        channel = this.getDmChannel(name);
+        if (!channel) {
+            channel = ChannelStore.getByName(name);
+        }
+
+        this.switchToChannel(channel);
+    }
+
+    getDmChannel(name) {
```

Comment:
this function is always returning null

My suggestion is use something like this


```javascript
getDmChannel(name) {
        return new Promise((resolve) => {
            if (name.indexOf(Utils.localizeMessage('channel_switch_modal.dm', '(Direct Message)')) > 0) {
                const dmUsername = name.substr(0, name.indexOf(Utils.localizeMessage('channel_switch_modal.dm', '(Direct Message)')) - 1);
                const user = UserStore.getProfileByUsername(dmUsername);

                if (user) {
                    openDirectChannelToUser(
                        user,
                        (ch) => {
                            resolve(ch);
                        },
                        null
                    );
                }
            }
            resolve(null);
        });
    }
```

and then use something like this

```javascript
handleSubmit(evt) {
        evt.preventDefault();

        const name = this.state.text.trim();
        this.getDmChannel(name).then((ch) => {
            let channel = ch;
            if (!channel) {
                channel = ChannelStore.getByName(name);
            }

            this.switchToChannel(channel);
        });
    }
```